feat(analysis/calculus/mean_value): more corollaries of the MVT

urkud opened the pull request:
TO CONTRIBUTORS:

Make sure you have:

  * [X] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [X] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [X] make sure definitions and lemmas are put in the right files
  * [X] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

urkud committed:
feat(analysis/calculus/mean_value): more corolaries of the MVT

urkud committed:
Fix compile, add "strict inequalities" versions of some theorems, add docs

urkud committed:
Merge branch 'master' into convex-on-deriv

sgouezel commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -221,3 +234,149 @@ exists_has_deriv_at_eq_slope f (deriv f) hab hfc
+  ∀ x y ∈ D, x < y → C * (y - x) < f y - f x :=
+begin
+  assume x y hx hy hxy,
+  -- cases eq_or_lt_of_le hxy with hxy' hxy', by rw [hxy', sub_self, sub_self, mul_zero],
```
Remove this comment?

urkud replied:
Done.

sgouezel commented on src/analysis/convex.lean:
```diff
@@ -412,18 +412,51 @@ begin
-lemma convex_on_linorder [linear_order α] {f : α → ℝ} : convex_on D f ↔
-  convex D ∧ ∀ {x y : α} {a b : ℝ}, x ∈ D → y ∈ D → x < y → 0 ≤ a → 0 ≤ b → a + b = 1 →
-    f (a • x + b • y) ≤ a * f x + b * f y :=
+/-- For functions on linear ordered sets, it suffices to verify the requirements for `x < y`
```
This docstring can not be understood just by itself. You say "it suffices", but you don't say what you want to do. Add "to prove convexity" just before "it suffices", for instance, to make it self-containted.

urkud replied:
Tried to fix.

sgouezel commented on src/analysis/convex.lean:
```diff
@@ -412,18 +412,51 @@ begin
+    exact hf hx hy hxy ha hb hab }
+end
+
+/-- This way of proving convexity of a function is used in the proof
```
The docstring should also say what the lemma does, not just where it is used.

urkud replied:
See new version.

sgouezel reviewed (commented):
Would it be worth adding versions of your statements with `D = univ` (replacing `differentiable_on` with `differentiable`, and removing `continuous_on` as this follows from differentiability)?

urkud commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -221,3 +234,149 @@ exists_has_deriv_at_eq_slope f (deriv f) hab hfc
+  ∀ x y ∈ D, x < y → C * (y - x) < f y - f x :=
+begin
+  assume x y hx hy hxy,
+  -- cases eq_or_lt_of_le hxy with hxy' hxy', by rw [hxy', sub_self, sub_self, mul_zero],
```
```suggestion
```

urkud committed:
Update src/analysis/calculus/mean_value.lean

urkud committed:
Merge branch 'master' into convex-on-deriv

urkud committed:
Add theorems for `convex_on univ`

sgouezel commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -221,3 +234,167 @@ exists_has_deriv_at_eq_slope f (deriv f) hab hfc
+  ∀ x y ∈ D, x ≤ y → f y ≤ f x :=
+by simpa only [zero_mul, sub_nonpos] using hD.image_sub_le_mul_sub_of_deriv_le hf hf' hf'_nonpos
+
+/-- If a function `f` is continuous on convex set `D ⊆ ℝ`, is differentiable on its interior,
```
```suggestion
/-- If a function `f` is continuous on a convex set `D ⊆ ℝ`, is differentiable on its interior,
```

sgouezel commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -221,3 +234,167 @@ exists_has_deriv_at_eq_slope f (deriv f) hab hfc
+  exact hf'_mono a b (hxyD' ⟨hxa, hay⟩) (hyzD' ⟨hyb, hbz⟩) (le_of_lt $ lt_trans hay hyb)
+end
+
+/-- If a function `f` is continuous on convex set `D ⊆ ℝ`, is differentiable on its interior,
```
```suggestion
/-- If a function `f` is continuous on a convex set `D ⊆ ℝ`, is differentiable on its interior,
```

sgouezel commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -221,3 +234,167 @@ exists_has_deriv_at_eq_slope f (deriv f) hab hfc
+convex_on_of_deriv_mono convex_univ hf.continuous.continuous_on hf.differentiable_on
+  (λ x y _ _ h, hf'_mono h)
+
+/-- If a function `f` is continuous on convex set `D ⊆ ℝ`, is twice differentiable on its interior,
```
```suggestion
/-- If a function `f` is continuous on a convex set `D ⊆ ℝ`, is twice differentiable on its interior,
```

urkud committed:
Fix comments

urkud committed:
@sgouezel adds missing articles

Thanks a lot! We don't have them in Russian, so it's hard for me to put them right.

Co-Authored-By: sgouezel <sebastien.gouezel@univ-rennes1.fr>

sgouezel commented on src/analysis/calculus/mean_value.lean:
```diff
@@ -221,3 +234,167 @@ exists_has_deriv_at_eq_slope f (deriv f) hab hfc
+end
+
+/-- Let `f` be a function continuous on a convex (or, equivalently, connected) subset `D`
+of the real line. If `f` is differentiable on the interior of `D` and `f' C C`, then
```
```suggestion
of the real line. If `f` is differentiable on the interior of `D` and `f' < C`, then
```

sgouezel reviewed (commented):
I still think it is worth adding versions of `convex.image_sub_lt_mul_sub_of_deriv_lt` and all the other theorems for `D = univ` (just removing `convex` from the name).

urkud committed:
Update src/analysis/calculus/mean_value.lean

Co-Authored-By: sgouezel <sebastien.gouezel@univ-rennes1.fr>

urkud committed:
Add more `univ` versions

sgouezel reviewed (approved):
Thanks!

mergify[bot] committed:
Merge branch 'master' into convex-on-deriv
